### PR TITLE
8317356: Fix missing null checks in the ClassFile API

### PR DESCRIPTION
**Pull request by liach**
Please review this patch that adds a test by @nizarbenalla to perform null checks across the ClassFile API. This is an updated version of #20556 that minimizes impact on our implementation code.

Notes:
1. There's one change in `MethodHandleProxies` to explicitly use platform class loader instead of `null` for boot class loader. Tests work fine.
2. The null check test uses the same set of classes as CorpusTest to make it suitable for tier 1.
3. The test is not comprehensive; currently it is permissive toward IAE, because sometimes some substitution arguments are invalid, especially primitive ClassDesc or Opcode with wrong types.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8317356](https://bugs.openjdk.org/browse/JDK-8317356): Fix missing null checks in the ClassFile API (**Enhancement** - P4)

### Reviewers
 * [Adam Sotona](https://openjdk.org/census#asotona) (@asotona - **Reviewer**)

### Contributors
 * Nizar Benalla `<nbenalla@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21458/head:pull/21458` \
`$ git checkout pull/21458`

Update a local copy of the PR: \
`$ git checkout pull/21458` \
`$ git pull https://git.openjdk.org/jdk.git pull/21458/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21458`

View PR using the GUI difftool: \
`$ git pr show -t 21458`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21458.diff">https://git.openjdk.org/jdk/pull/21458.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21458#issuecomment-2405861612)

### CODE REVIEW

File: test/jdk/jdk/classfile/NullHostileTest.java
Comment (asotona) [inline test/jdk/jdk/classfile/NullHostileTest.java:1]: I'm a bit worried about maintainability of such complex parametrized test with hard-listed method signatures in a text form. I would recommend to find better location/category and tier for the test.

Comment (ExE-Boss) [reply test/jdk/jdk/classfile/NullHostileTest.java:1]: Perhaps the parameters could be annotated with a JDK internal `@Nullable` annotation, which would then be used by the test to determine whether the parameter is supposed to allow `null`s.

--------------------------------------------------------------------------------

This would also ease future migration to [Null‑Restricted and Nullable Types].

[Null‑Restricted and Nullable Types]: https://openjdk.org/jeps/8303099

Comment (liach) [reply test/jdk/jdk/classfile/NullHostileTest.java:1]: Should I remove this test and only include the null checks in the pr?

File: src/java.base/share/classes/java/lang/classfile/instruction/DiscontinuedInstruction.java [hunk 1]
```diff
@@ -34,6 +34,8 @@
 import jdk.internal.classfile.impl.Util;
 import jdk.internal.javac.PreviewFeature;
 
+import static java.util.Objects.requireNonNull;
```
Comment (asotona) [inline src/java.base/share/classes/java/lang/classfile/instruction/DiscontinuedInstruction.java:37]: Isn't it unused?

File: src/java.base/share/classes/jdk/internal/classfile/impl/ClassRemapperImpl.java [hunk 2]
```diff
@@ -273,7 +273,6 @@ public CodeTransform asCodeTransform() {
 
     @Override
     public ClassDesc map(ClassDesc desc) {
-        if (desc == null) return null;
```
Comment (asotona) [inline src/java.base/share/classes/jdk/internal/classfile/impl/ClassRemapperImpl.java:276]: I'm not sure this won't break some use case.
Here I would say null -> null mapping might be OK.

Comment (liach) [reply src/java.base/share/classes/jdk/internal/classfile/impl/ClassRemapperImpl.java:276]: Makes sense. I will push the changes when you finish the review.

Comment (asotona) [reply src/java.base/share/classes/jdk/internal/classfile/impl/ClassRemapperImpl.java:276]: It is all I have.

**Comment by liach** [comment]
Thanks adam for the review. I have removed the test for now. We can come back later and find a more comprehensive way to test builders accessible only in transforms, etc. and the null check test might actually be incorporated with a general contract check test that checks for IAE too.

**Review by asotona — APPROVED** [review]
Looks good to me.

**Comment by ExE-Boss** [comment]
Since this PR no longer adds a test, the issue should probably be renamed.

**Pull request closed without merging**